torch/testing/_internal/common_utils.py

```diff
@@ -1072,24 +1072,112 @@ def setUp(self):
 
         set_rng_seed(SEED)
 
+    @staticmethod
+    def _make_crow_indices(n_rows, n_cols, nnz,
```

Comment:
I honestly have to admit that the following code isn't obvious to me. I took a look at the section in your [blog post](https://pearu.github.io/csr_sampling.html#new-algorithm), but I'm having a hard time understanding how the histogram yields counts that satisfy the conditions you mention and then how you're parameterizing the space of counts that yield it via below algorithm. For long term maintainability, could I ask you to add a few more comments to each part of the algorithm below?

Reply:
@cpuhrsch I have added the following description of the method to _make_crow_indices doc-string:
```
        Sampling method
        ---------------

        The used sampling method was introduced in
        https://pearu.github.io/csr_sampling.html, and here we give
        only an overall description of the method.

        Notice that crow_indices can be defined as cumsum(counts)
        where counts is a sequence of non-negative integers satisfying
        the following conditions:

          len(counts) == n_rows + 1
          counts.max() <= n_cols

        while counts[i + 1] is interpreted as the number of specified
        elements in the i-th row.

        The used sampling method aims at increasing the diversity of
        CSR samples, that is, a CSR sample should contain (i) rows
        that are all filled, (ii) rows with no elements at all, and
        (iii) rows that are partially filled. At the same time and for
        the given total number of specified elements (nnz), there
        should be minimal preference to rows with a given number of
        elements.  To achieve this, the sampling method is built-up on
        using a sawteeth model for counts. In the simplest case, we
        would have

          counts = arange(n_rows + 1) % (n_cols + 1)

        that has equal number of all possible column counts per row.
        This formula can be used only for specific input values of
        n_rows, n_cols, and nnz. To generalize this model to any
        combinations of inputs, the counts model above is extended
        with an incomplete sawtooth, and the right and lower
        rectangular parts that will guarantee that

          counts.sum() == nnz

        for any combination of n_rows, n_cols, and nnz.
```
I hope it clarifies how the algorithm works in general.

In addition, I'll try to comment on the implementation of the algorithm as well to make it clear what each part does. However, proving all the formula in detail might not help either as the description would become too long. That is compensated by
https://github.com/pytorch/pytorch/pull/60283/files#diff-ee9917f2665422eb0840197fb632f86b6c69b0e7a618123852de337038e5e3a1R24-R34
that tests all combinations of inputs (for reasonable sizes) against the requirements of `crow_indices`.